pkg/server/status.go

```diff
@@ -1989,6 +1990,104 @@ func (s *statusServer) NodeUI(
 	return &resp, nil
 }
 
+// NetworkConnectivity collects information about connections statuses across all nodes.
+func (s *systemStatusServer) NetworkConnectivity(
+	ctx context.Context, req *serverpb.NetworkConnectivityRequest,
+) (*serverpb.NetworkConnectivityResponse, error) {
+	ctx = forwardSQLIdentityThroughRPCCalls(ctx)
+	ctx = s.AnnotateCtx(ctx)
+
+	response := &serverpb.NetworkConnectivityResponse{
+		Connections:    map[roachpb.NodeID]serverpb.NetworkConnectivityResponse_Connectivity{},
+		ErrorsByNodeID: map[roachpb.NodeID]string{},
+	}
+
+	if len(req.NodeID) > 0 {
+		sourceNodeID, local, err := s.parseNodeID(req.NodeID)
+		if err != nil {
+			return nil, serverError(ctx, err)
+		}
+
+		if local {
+			peers := map[roachpb.NodeID]serverpb.NetworkConnectivityResponse_Peer{}
+			var nodeIDs []roachpb.NodeID
+			err = s.gossip.IterateInfos(gossip.KeyNodeDescPrefix, func(k string, info gossip.Info) error {
+				nodeIDs = append(nodeIDs, info.NodeID)
+				return nil
+			})
+			if err != nil {
+				return nil, serverError(ctx, err)
+			}
+
+			latencies := s.rpcCtx.RemoteClocks.AllLatencies()
+
+			for _, targetNodeId := range nodeIDs {
+				if sourceNodeID == targetNodeId {
+					continue
+				}
+				peer := serverpb.NetworkConnectivityResponse_Peer{}
+				peer.Latency = latencies[targetNodeId]
+
+				node, err := s.gossip.GetNodeDescriptor(targetNodeId)
+				if err != nil {
+					peer.Status = serverpb.NetworkConnectivityResponse_UNKNOWN
+					continue
+				}
+				if err = s.rpcCtx.ConnHealth(node.Address.AddressField, node.NodeID, rpc.SystemClass); err != nil {
+					if kvpb.IsDecommissionedStatusErr(err) || errors.Is(rpc.ErrQuiescing, err) || grpcutil.IsClosedConnection(err) {
+						peer.Status = serverpb.NetworkConnectivityResponse_CLOSED
+					} else if errors.Is(rpc.ErrNotHeartbeated, err) {
+						peer.Status = serverpb.NetworkConnectivityResponse_NOT_CONNECTED
+					} else {
+						peer.Status = serverpb.NetworkConnectivityResponse_PARTITIONED
```

Comment:
This one would go away. I don't think we are in the right place here to reason about what a connection error "means"; the code should avoid trying to do that (it's not even something we can do in general). The connection is either absent, initializing, errored, or established. Exposing the error is more important than trying to classify it.

Reply:
>  The connection is either absent, initializing, errored, or established.

@tbg , technically all connections are closed with errors, but would it make sense to indicate connections that were closed intentionally as errored? I'd try to distinguish those connections.

Reply:
I'm not sure, but it would also require another update to the `rpc` package. We can consider it, but let's not hold up this work for it.

If you are exposing the error text as I suggested elsewhere, the error message can also be a carried for this, so we may not need a new enum state. But we basically never close connections on purpose anyway.

Reply:
Done. Updated connection statuses to rely on `ERROR` status and provide error info whenever this information is available.